### ISSUE TITLE
Identify the anomaly fixed in Luminary 99 Rev 1

### ISSUE THREAD

**Issue by thewonderidiot** [virtualagc/virtualagc#1043]
We've recently discovered that Luminary 99R1 was created last-minute to fix an anomaly discovered in Luminary 99, which had been around since at least Luminary 69. The fix, simply put, makes ENEMA reset timers 3, 4, and 5.

The branch of the LEMONAID subroutine created to correct the problem was entitled `LNYAIDE`, which strongly suggests that this fix was associated with an LNY- numbered anomaly.

Thanks to [Luminary Memo 91](http://www.ibiblio.org/apollo/Documents/LUM91_text.pdf), we have a list of all anomalies fixed in Luminary from revision 100 to 105. [Luminary Memo 96](http://www.ibiblio.org/apollo/Documents/LUM96_text.pdf), the Level 3 Test Plan for Luminary 1B, gives short text descriptions of all of them except for LNY-65 (but we know about that one from [Luminary Memo 83](http://www.ibiblio.org/apollo/Documents/LUM83_text.pdf)). Memo 91 was published July 7, and Luminary 99R1 was released sometime in June, so we know that the fix had to have gone into mainline Luminary no later than revision 105. It is also quite likely that this anomaly had not been known about for long. Since Luminary revision 99 was created to patch LNY-70, I would hazard a guess that the 99R1 anomaly would be LNY-71 or higher, but that is not definite.

With this information, we should (hopefully) be able to identify exactly which anomaly 99R1 was created to fix. This is, of course, little more than historical curiosity, but it sure would be interesting to know!

The anomalies fixed in 100-105 are as follows:

| Anomaly | Fixed In | Description | Candidate |
| ------------ | ----------- | --------------- | ---------------------- |
| ~~LNY-58~~ | ~~100~~ | ~~It is possible to enter a program which uses the RR or LR while R12 is reading the LR.~~ | No (change to R12) |
| LNY-61 | 100 | If a restart occurs during the trunnion or shaft incorporations of P20 Rend. Nav., an incorrect State Vector Update can occur. | Yes |
| LNY-62 | 100 | Abort programs do not alter current autopilot deadband. | Yes |
| LNY-63 | 100 | DPS Engine Failure in P40. | Yes |
| LNY-64 | 100 | Restart occurring in P63. | Yes |
| ~~LNY-65~~ | ~~98/100~~ | ~~The coding to initialize the LAMBERT iteration counter (ITERCTR) was restored to INITVEL.~~ | No (fixed in 98, code moved to correct spot in 100) |
| ~~LNY-66~~ | ~~100~~ | ~~R04 Erasable Conflict with R65~~ | No (moved erasables) |
| LNY-67 | 100 | Make ATTSTALL 1210 into BAILOUT instead of POODOO. | Yes |
| LNY-69 | 101 | Eliminate Coding error causing downward jet firing. | Yes |
| ~~LNY-71~~ | ~~100~~ | ~~In P22 Lunar Surface Nav. the RR predesignate routine stores into erasable used by orbital integration.~~ | No (predesignate routine and/or erasable fix) |
| LNY-72 | 100 | Program alarm light every 2 sec. after V32E to 523 alarm. | Yes |
| ~~LNY-73~~ | ~~100~~ | ~~Bug in recycle option to N61 in P63~~ | No (bug in P63) |
| ~~LNY-74~~ | ~~100~~ | ~~Erasable Conflict when LOCSAM (i. e. , LSPOS)~~ | No (moved erasables) |
| ~~LNY-76~~ | ~~102~~ | ~~N60 Made to be HCALC1 (ALTRATE)~~ | No (N60 change) |
| LNY-77 | 102 | No takeover when Abort from P60's | Yes |

**Comment by rburkey2005** [comment]
I must have learned something from the bosses I've encountered who insisted on offering completely unstudied opinions on the basis of their gut impressions.  :smile:  LNY-61 looks suspicious.

**Comment by thewonderidiot** [comment]
Niklas and I are on the same page. That one is a bit troublesome to test, since the incorporations are quite fast, so if using V69 to trigger a TC Trap turns out to be too hard we've got schemes force restarts other ways. I think he's going to test that at some point, between his work on the AGS and flight plan for his Zerlina mission, while I trace through the possible routes to ENEMA and study diffs between Luminary 99 and 116 and try to associate code changes with anomalies or PCRs.

**Comment by thewonderidiot** [comment]
Three and a half years later, and I've finally figured it out! The anomaly is **LNY-77** and it's actually very interesting. In addition to the title "No takeover when Abort from P60's" mentioned in memo 96, we have just a few other pieces of information about it.

First, this snapshot of a slide for the Luminary 1B Flight Software Readiness Review:
![image](https://user-images.githubusercontent.com/94056/107082039-74ff7f00-67a8-11eb-9eea-57ca8c06e2f4.png)
On it, Russ Larson has penciled in "77  Software restart will not now cause up to 163 sec activity delay". This handwritten note is what made me originally think LNY-77 was most likely it -- since the anomaly was technically fixed in _both_ Luminary 1A and 1B, it seems reasonable that it would have been initially excluded from the list of changes.

Russ's description of the anomaly led me to another snapshot I have from his papers:
![image](https://user-images.githubusercontent.com/94056/107082471-125ab300-67a9-11eb-8788-039cfd774cfe.png)

This appears to be a draft addition to the Luminary 1A program notes, and while it doesn't explicitly mention the name LNY-77 anywhere, it matches Russ's description quite closely ("There is a very low probability that the LGC might stop the navigation and guidance processors for about 160 seconds...").

After lots of "what ifs" I was able to reproduce the bug, and the way it happens is very interesting. It's actually a failure of `WAITLIST`, due to a few of its specific implementation details, and how the P70/P71 selection is accomplished.

In normal operation, `WAITLIST` schedules TIME3 such that the next `T3RUPT` occurs at the moment the next task is supposed to run. It maintains a list of upcoming task addresses in `LST2`, and a list of times to wait between those tasks in `LST1`. When `WAITLIST` adds tasks to the lists, it calculates the new `LST1` entry as `1 - (delta time between this task and next)`:
![image](https://user-images.githubusercontent.com/94056/107085268-15f03900-67ad-11eb-99ea-869afe8ea23d.png)

Before kicking off the task, `T3RUPT` does a little bit of internal management on these lists. It pops the current task's address and new delta-T off of the front of the queues, and then calculates the new value to set TIME3.
![image](https://user-images.githubusercontent.com/94056/107085345-38825200-67ad-11eb-90a2-45fc5dda47ef.png)

`TIME3` is expected to be 0 here, since its rollover is what has just triggered the `T3RUPT`. So the new TIME3 calculation ends up being
```
TIME3 = 16383 + (1 - deltaT)
    = 16384 - deltaT
```
Note that if deltaT is 0 (i.e. two tasks were scheduled for the same `T3RUPT`), this calculation will result in a positive overflow. In case of an overflow, as shown in the pseudocode above, the variable `RUPTAGN` gets set to 1, and TIME3 is left as-is -- again equal to `00000` and counting up.

Because of this, all `WAITLIST` tasks are required to call `TASKOVER` upon completion. This routine checks the state of `RUPTAGN`, and if it is set, jumps back to the start of `T3RUPT` processing, where the next values will be popped off o f the lists, and `TIME3` recalculated. If `RUPTAGN` is _not_ 1, `TASKOVER` simply RESUMEs to the interrupted location.
![image](https://user-images.githubusercontent.com/94056/107085915-02919d80-67ae-11eb-971c-6c363abed806.png)

The ABORT and ABORT STAGE buttons are monitored via the WAITLIST task `R10,R11`, which runs once every 250ms throughout powered descent, starting up alongside the `SERVICER` at T-35s in P63. If it detects either button down, it checks legality of the selection, and then transfers control to `CONTABRT`.
```
CONTABRT	CAF	ABRTJADR
		TS	BRUPT
		RESUME

ABRTJADR	TCF	ABRTJASK

ABRTJASK	CAF	OCTAL27
....
```
`CONTABRT` replaces the interrupted instruction in `BRUPT` with a jump to `ABRTJASK`, and then exits the `T3RUPT` interrupt directly to it. `ABRTJASK` does a lot of flag manipulation, but its most important function is setting up the restart tables to actually start up the selected abort program.
![image](https://user-images.githubusercontent.com/94056/107087403-3c63a380-67b0-11eb-91a4-df8fb7ae4f48.png)
Specifically, it sets up `4.3SPOT` for `GOABORT`, `5.4SPOT` for `SERVICER`, and `2.21SPOT` for `R10,R11`. It then calls `ENEMA`, the same software restart routine used by `BAILOUT` for e.g. 1201 or 1202 alarms.

`ENEMA` completely flushes out the system, and all of the currently running tasks and jobs get killed. The relevant restart table entries are then processed:
```
4.3SPOT		OCT	25000
		EBANK=	DVCNTR
		2CADR	GOABORT
```
```
5.4SPOT		DEC	200
		EBANK=	DVCNTR
		-2CADR	REREADAC

		OCT	20000
		EBANK=	DVCNTR
		2CADR	SERVICER
```
```
2.21SPOT	DEC	25
		EBANK=	DVCNTR
		-2CADR	R10,R11
```
And with all of these started back up, the system proceeds with the abort guidance.

All of this is fine in 99% of situations. Imagine, though, the very rare case where the `R10,R11` that detects one of the abort buttons down had another task scheduled after it for the same `T3RUPT`. In this case, the abort code completely bypasses the logic and `TASKOVER`, and by the time control reaches the `ENEMA`, `TIME3` is still 0, and `RUPTAGN` is 1.

The restarts routine uses dynamically-constructed calls to `NOVAC`, `FINDVAC`, or `WAITLIST` based on the contents of the restart tables to relaunch all of the desired jobs and tasks. So, based on the above restart groups, `SERVICER` and `GOABORT` are kicked off with `FINDVAC`, while `R10,R11` and `REREADAC` are both scheduled with calls to `WAITLIST`.

`WAITLIST`, on scheduling a new task, first tries to determine (a) if the new task is to be scheduled sooner than the task on the front of the list, in which case `TIME3` is set to the required deltaT, and (b) if `TIME3` is a small positive number, less than 128, in which case the code assumes that `TIME3` has just rolled over and `T3RUPT` is actively pending. In case of (b), it completely skips setting of `TIME3` and simply adds the new tasks to the lists.

In our scenario, however, `T3RUPT` is _not_ pending, and `TIME3` is certainly less than 128, so nothing actually ever resets it. `SERVICER` will run once and exit, and nothing else will happen until `TIME3` counts all the way up to `37777` from `00000`, which takes 163 seconds.

The fix for this then is nice and clean: make `ENEMA` reset `TIME3` to a sane value. This was easily done by moving the `STARTSB1` label up a few instructions -- which is exactly the change we see between 99 and 99/1.

![image](https://user-images.githubusercontent.com/94056/107092643-d3ccf480-67b8-11eb-9bbc-f12a7e63c052.png)

Note that while the anomaly title and the program note only explicitly mention the ABORT and ABORT STAGE buttons, Russ Larson's note generalizes it to "software restarts". The key requirement for this anomaly to trigger is that a waitlist task that has another task scheduled to follow it in the same T3RUPT instead causes a software restart via `ENEMA`.

This is _particularly_ relevant for the Apollo 11 landing, because each of the `1201` and `1202` alarms happened when trying to spawn jobs (be they `SERVICER` or monitor verb updates) from waitlist tasks. `1201` and `1202` are both `BAILOUT`-type alarms -- and `BAILOUT` exits via a jump to `ENEMA`. Were LNY-77 triggered in this fashion, `SERVICER` would exit for 163 seconds during the landing... rendering the AGC useless, and necessitating an AGS abort. So while it's impossible to know, there is a very real chance that the LNY-77 bugfix might have saved the Apollo 11 landing.

Here's a gif of LNY-77 in action in VirtualAGC. In it, I'm simulating a press of the ABORT STAGE button by manually changing the relevant bit in channel 30.
![lny77_demo](https://user-images.githubusercontent.com/94056/107093289-d4b25600-67b9-11eb-88af-7db6afb5e75a.gif)

Because the circumstances under which the anomaly occurs are quite rare, I've written an EMP that guarantees the bug will be triggered, by scheduling itself over top of all instances of `R10,R11`:
```
 0660    00000            OCT    0
 0661    34752   FINDSERV CAF    TWO                 ; Wait for 20ms before checking
 0662    05224            TC     VARDELAY
 0663    41410            CS     LST2                ; Is READACCS the next scheduled task?
 0664    60667            AD     RDACCSAD
 0665    10000            CCS    A
 0666    00661            TC     FINDSERV            ; No, check again in 20ms
 0667    03441   RDACCSAD REMADR READACCS            ; Unused CCS branch; used for READACCS address
 0670    00661            TC     FINDSERV            ; No, check again in 20ms
 0671    40026   FOUNDIT  CS     TIME3               ; Yes. Calculate how far in the future it is going
 0672    64736            AD     HALF                ; to execute, and schedule ourself on top of it.
 0673    64736            AD     HALF
 0674    05224   AFTSERV  TC     VARDELAY
 0675    36007            CAF    OCT24               ; We are now running in the same T3RUPT as READACCS.
 0676    05224   R10SYNC  TC     VARDELAY            ; Sleep for 200ms to schedule on top of R10,R11.
 0677    11056            CCS    PIPCTR              ; Look at PIPCTR to determine R10,R11's next move.
 0700    00712            TC     STILLR10            ; PIPCTR > 0. R10,R11 rescheduled itself for 250ms.
 0701    36010            CAF    OCT31               ; PIPCTR = 0. R10,R11 will exit after its next 250ms wakeup.
 0702    05224            TC     VARDELAY
 0703    00710            TC     +5
 0704            GOLOC -1                            ; These four locations are used by the RESTARTS ROUTINE.
 0705            GOLOC
 0706            GOLOC +1
 0707            GOLOC +2
 0710    34756            CAF    FIVE                ; R10,R11 has exited. Wake up in 50ms to meet READACCS again.
 0711    00674            TC     AFTSERV  
 0712    36010   STILLR10 CAF    OCT31               ; Go wake up on top of R10,R11 in 250ms.
 0713    00676            TC     R10SYNC 
```
The EMP should be loaded into VAC area 5 before P63 is started. After P63 is selected, the EMP is started via:
```
V21N01E 660E 0E       ; Reclaim VAC 5
V25N26E 2E 661E 3E    ; Set up waitlist task starting in 20ms at address 661, using EBank 3
V31E                  ; Begin task
```
After this is done, LNY-77 is guaranteed to be triggered on an abort or software restart.

**Comment by rburkey2005** [comment]
Congratulations!  The writeup is amazing to read ... though I'm not amazed that you eventually cracked it.  (I'm most amazed to see that I apparently had felt called upon to [write a comment](#issuecomment-321666442) offering an unhelpful, uninformed opinion about a different LNY.  Neither do I have any idea why I thought LNY-61 was "suspicious", nor any memory at all of offering an opinion about it.  Just like the very bosses I said in my earlier comment that I was probably emulating.  The circle is now complete; the student &mdash; of unhelpful opinions &mdash; has become the master. :smile:)

**Comment by indy91** [comment]
It was a bit of a hassle to get the EMP working in NASSP, but after many tries and revisions of the EMP we finally got there. My favourite variation of the bug is when I do an early abort during the descent. In some cases Noun 63 is shown in P70, but only once and the DSKY never updates in any way after that. And it also has already commanded the thrust vector to point forward instead of backwards to get back into orbit, so early in the descent that is a 180° pitch maneuver the LM has to do. But with the bug it never tries to stop at the right attitude and all that was accomplished is to send the LM in a pretty fast spin. My instinct definitely says to switch to the AGS at that point!
